fairscale/optim/oss.py

```diff
@@ -102,9 +122,41 @@ def step(self, closure: Optional[Callable[[], float]] = None, **kwargs: Any) ->
 
         # Sync all the states
         for rank, param_groups in enumerate(self.partition_parameters()):
+            # Batch smaller params in a broadcast buffer to speed up the communication
+            buffered_params: List[Parameter] = []
+            buffered_elements = 0
+
             for param_group in param_groups:
+                # We assume that the whole param group is on the same device
```

Comment:
This is not a safe assumption. gpipe we have params on multiple devices within the same group.

Reply:
ouch, yes of course, it's an old comment but I stuck to it in the logic, that's pretty bad. Sorry for the dumb submission, fixing that